File: gcs/ui/mode_rail.py
"""Left action/flight-mode rail — the vertical button strip on the fly-view.

Mirrors the NASA-style ground-station layout: a stack of big, glanceable
icon+label buttons down the left edge for the common in-flight actions (LAND,
RETURN, PAUSE) plus an ACTION button (take off), then a SINGLE/MULTI selector at
the bottom. The action buttons map to ArduCopter flight modes and light up when
that mode goes live (read back from HEARTBEAT); the view only announces intent
(``mode_requested`` / ``takeoff_requested``) and the MainWindow routes it to the
command service, so this stays a passive view.
"""
from __future__ import annotations

from PySide6.QtCore import Qt, Signal
from PySide6.QtWidgets import (QButtonGroup, QFrame, QPushButton, QSizePolicy,
                               QVBoxLayout)

from ..domain import flight_modes
from ..domain.telemetry import TelemetrySnapshot

# (glyph, label, mode-name) for the mode buttons — mode None means "take off".
_ACTIONS = [
    ("⬇", "LAND", "LAND"),
    ("↩", "RETURN", "RTL"),
    ("⏸", "PAUSE", "LOITER"),
    ("▶", "ACTION", None),
]


class ModeRail(QFrame):
    mode_requested = Signal(str)     # canonical mode name
    takeoff_requested = Signal()     # ACTION button — MainWindow prompts altitude

    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("ModeRail")
        self.setSizePolicy(QSizePolicy.Maximum, QSizePolicy.Maximum)
        self._connected = False
        self._mode_buttons: dict[str, QPushButton] = {}

        col = QVBoxLayout(self)
        col.setContentsMargins(6, 6, 6, 6)
        col.setSpacing(8)

        for glyph, label, mode in _ACTIONS:
            btn = self._make_button(glyph, label)
            if mode is None:
                btn.clicked.connect(self.takeoff_requested.emit)
                btn.setToolTip("Take off (arm + GUIDED, prompts for altitude)")
                self._action_btn = btn
            else:
                btn.setCheckable(True)
                btn.clicked.connect(lambda _=False, m=mode: self.mode_requested.emit(m))
                btn.setToolTip(f"Switch to {mode}")
                self._mode_buttons[mode] = btn
            col.addWidget(btn)

        col.addSpacing(6)

        # single / multi vehicle selector (cosmetic — this build flies one vehicle)
        self._single = self._make_button("◈", "SINGLE")
        self._single.setCheckable(True)
        self._single.setChecked(True)
        self._multi = self._make_button("⧉", "MULTI")
        self._multi.setCheckable(True)
        self._multi.setEnabled(False)
        self._multi.setToolTip("Multi-vehicle control is not available in this build")
        grp = QButtonGroup(self)
        grp.setExclusive(True)
        grp.addButton(self._single)
        grp.addButton(self._multi)
        col.addWidget(self._single)
        col.addWidget(self._multi)

        self.set_connected(False)

    @staticmethod
    def _make_button(glyph: str, label: str) -> QPushButton:
        btn = QPushButton(f"{glyph}\n{label}")
        btn.setObjectName("RailBtn")
        btn.setCursor(Qt.PointingHandCursor)
        btn.setFixedSize(62, 56)
        return btn

    # ── external state ────────────────────────────────────────────────────────
    def set_connected(self, connected: bool) -> None:
        self._connected = connected
        for btn in self._mode_buttons.values():
            btn.setEnabled(connected)
        self._action_btn.setEnabled(connected)
        if not connected:
            for btn in self._mode_buttons.values():
                btn.setChecked(False)

    def update_from(self, s: TelemetrySnapshot) -> None:
        active = (flight_modes.mode_name(s.mode.autopilot, s.mode.custom_mode)
                  if s.heartbeat_seen else None)
        for mode, btn in self._mode_buttons.items():
            btn.setChecked(mode == active)

File: cpp/src/ui/mode_rail.cpp
#include "ui/mode_rail.h"

#include "domain/flight_modes.h"
#include "ui/theme.h"

#include <QButtonGroup>
#include <QSizePolicy>
#include <QToolButton>
#include <QVBoxLayout>

namespace gcs::ui {

namespace {
struct Action { const char *icon; QString label; QString mode; bool isTakeoff; };
}

ModeRail::ModeRail(QWidget *parent) : QFrame(parent)
{
    setObjectName("ModeRail");
    setSizePolicy(QSizePolicy::Maximum, QSizePolicy::Maximum);

    // (biểu tượng, nhãn, tên-chế-độ) — mode rỗng nghĩa là "cất cánh".
    const std::vector<Action> actions = {
        {":/icons/land.png", QStringLiteral("HẠ CÁNH"), QStringLiteral("LAND"), false},
        {":/icons/return.png", QStringLiteral("VỀ"), QStringLiteral("RTL"), false},
        {":/icons/pause.png", QStringLiteral("TẠM DỪNG"), QStringLiteral("LOITER"), false},
        {":/icons/takeoff.png", QStringLiteral("CẤT CÁNH"), QString(), true},
    };

    auto *col = new QVBoxLayout(this);
    col->setContentsMargins(6, 6, 6, 6);
    col->setSpacing(8);

    for (const auto &a : actions) {
        auto *btn = makeButton(theme::railIcon(QString::fromUtf8(a.icon)), QString(), a.label);
        if (a.isTakeoff) {
            connect(btn, &QToolButton::clicked, this, &ModeRail::takeoffRequested);
            btn->setToolTip(QStringLiteral("Cất cánh (arm + GUIDED, hỏi độ cao)"));
            m_actionBtn = btn;
        } else {
            btn->setCheckable(true);
            const QString mode = a.mode;
            connect(btn, &QToolButton::clicked, this, [this, mode] { emit modeRequested(mode); });
            btn->setToolTip(QStringLiteral("Chuyển sang %1").arg(mode));
            m_modeButtons.insert(mode, btn);
        }
        col->addWidget(btn);
    }

    col->addSpacing(6);

    // bộ chọn đơn/đa phương tiện (chỉ để trang trí — bản này bay một phương tiện)
    m_single = makeButton(QIcon(), QStringLiteral("◈"), QStringLiteral("ĐƠN"));
    m_single->setCheckable(true);
    m_single->setChecked(true);
    m_multi = makeButton(QIcon(), QStringLiteral("⧉"), QStringLiteral("ĐA"));
    m_multi->setCheckable(true);
    m_multi->setEnabled(false);
    m_multi->setToolTip(QStringLiteral("Điều khiển đa phương tiện không có ở bản này"));
    auto *grp = new QButtonGroup(this);
    grp->setExclusive(true);
    grp->addButton(m_single);
    grp->addButton(m_multi);
    col->addWidget(m_single);
    col->addWidget(m_multi);

    setConnected(false);
}

QToolButton *ModeRail::makeButton(const QIcon &icon, const QString &glyph, const QString &label)
{
    auto *btn = new QToolButton;
    btn->setObjectName("RailBtn");
    btn->setToolButtonStyle(Qt::ToolButtonTextUnderIcon);
    btn->setCursor(Qt::PointingHandCursor);
    btn->setFixedSize(62, 56);
    if (!icon.isNull()) {
        btn->setIcon(icon);
        btn->setIconSize(QSize(24, 24));
        btn->setText(label);
    } else {
        // không có PNG: dùng glyph text trên nhãn
        btn->setText(glyph + QStringLiteral("\n") + label);
    }
    return btn;
}

void ModeRail::setConnected(bool connected)
{
    m_connected = connected;
    for (auto *btn : m_modeButtons)
        btn->setEnabled(connected);
    m_actionBtn->setEnabled(connected);
    if (!connected)
        for (auto *btn : m_modeButtons)
            btn->setChecked(false);
}

void ModeRail::updateFrom(const domain::TelemetrySnapshot &s)
{
    const QString active = s.heartbeatSeen
        ? domain::flight_modes::modeName(s.mode.autopilot, s.mode.customMode)
        : QString();
    for (auto it = m_modeButtons.constBegin(); it != m_modeButtons.constEnd(); ++it)
        it.value()->setChecked(it.key() == active);
}

} // namespace gcs::ui
